plugin.go

```diff
@@ -28,40 +45,259 @@ func (p *Plugin) PluginInfo() launchr.PluginInfo {
 
 // OnAppInit implements launchr.Plugin interface.
 func (p *Plugin) OnAppInit(app launchr.App) error {
+	app.GetService(&p.cfg)
+
 	p.app = app
 	return nil
 }
 
 // CobraAddCommands implements launchr.CobraPlugin interface to provide web functionality.
 func (p *Plugin) CobraAddCommands(rootCmd *cobra.Command) error {
 	// Flag options.
-	var port string
+	var port int
 	var proxyClient string
 	var useSwaggerUI bool
+	var foreground bool
 	var cmd = &cobra.Command{
-		Use:   "web",
-		Short: "Starts web server",
+		Use:       "web",
+		Short:     "Starts web server",
+		Args:      cobra.MatchAll(cobra.RangeArgs(0, 1), cobra.OnlyValidArgs),
+		ValidArgs: []string{stopArg},
 		RunE: func(cmd *cobra.Command, args []string) error {
 			// Don't show usage help on a runtime error.
 			cmd.SilenceUsage = true
 
+			pluginTmpDir := p.getPluginTempDir()
+			pidFile := fmt.Sprintf("%s/%s.pid", pluginTmpDir, strings.ToLower(cmd.Name()))
+
+			if !isBackGroundEnv() {
+				runInfo, err := server.GetRunInfo(pluginTmpDir)
+				if err != nil {
+					return err
+				}
+
+				// If 'stop' arg passed, try to interrupt process and remove PID file.
+				if len(args) > 0 && args[0] == stopArg {
+					err = stopWeb(pidFile, runInfo)
+					if err != nil {
+						return err
+					}
+
+					cli.Println("Web server stopped successfully")
+					return nil
+				}
+
+				if isWebRunning(pidFile, runInfo) {
+					return fmt.Errorf("server is already running by address %s, please stop it before starting new one", runInfo.BaseURL)
+				}
+			} else {
+				// Store background outputs into plugin tmp folder.
+				err := redirectOutputs(pluginTmpDir)
+				if err != nil {
+					return err
+				}
+			}
+
+			if !server.IsPortFree(port) {
+				log.Info("web server port %d you are trying to use is not available", port)
+				port = server.GetFreePort(port, 20)
+			}
+
 			runOpts := &server.RunOptions{
-				Addr:        fmt.Sprintf(":%s", port), // @todo use proper addr
-				APIPrefix:   APIPrefix,
-				SwaggerJSON: useSwaggerUI,
-				ProxyClient: proxyClient,
+				Addr:          fmt.Sprintf(":%d", port), // @todo use proper addr
+				APIPrefix:     APIPrefix,
+				SwaggerJSON:   useSwaggerUI,
+				ProxyClient:   proxyClient,
+				RunInfoFolder: pluginTmpDir,
 				// @todo use embed fs for client or provide path ?
 			}
-
 			prepareRunOption(p, runOpts)
 
+			if foreground {
+				return server.Run(cmd.Context(), p.app, runOpts)
+			}
+
+			if !isBackGroundEnv() {
```

Comment:
The function declaration is too fat. Try to refactor to meaningful functions.
If possible, move out of plugin the whole process management stuff. Concentrate on main behavior in the main file.